nixos/modules/services/security/step-ca.nix

```diff
@@ -0,0 +1,121 @@
+        description = "Hostname/FQDN of your CA.";
+      };
+      address = lib.mkOption {
+        type = lib.types.str;
+        default = "127.0.0.1";
+        example = "0.0.0.0";
+        description = "The address the CA will listen at.";
+      };
+      port = lib.mkOption {
+        type = lib.types.port;
+        default = 8443;
+        description = "Port to listen on.";
+      };
+      passwordFile = lib.mkOption {
+        type = lib.types.path;
+        default = null;
+        example = "/run/keys/smallstep-password";
+        description = "Path to file containing password of your first provisioner.";
+      };
+      provisioner = lib.mkOption {
+        type = lib.types.str;
+        default = "you@smallstep.com";
+        description = "The name of the first provisioner.";
+      };
+      acmeProvisioner = lib.mkOption {
+        type = lib.types.nullOr lib.types.str;
+        default = null;
+        example = "acme";
+        description = "Optionally create an acme provider. The resulting acme url is https://{hostname}:{port}/acme/{acmeProvisioner}/directory";
+      };
+      rootFile = lib.mkOption {
+        type = lib.types.nullOr lib.types.path;
+        default = null;
+        description = ''
+          The path of an existing PEM file to be used as the root certificate authority.
+          You can generate one using e.g `step certificate create "MyCompany" root-ca.crt root-ca.key --profile root-ca --no-password --insecure` from the `step-cli` package.
+          Only use this for development as the ca key is not protected with a password.
+          It seems like step-ca ignores the passwordFile parameter so you can't use a key with a password.
+          This should probably be reported as a bug. https://github.com/smallstep/cli/issues/220'';
+      };
+      keyFile = lib.mkOption {
+        type = lib.types.nullOr lib.types.path;
+        default = null;
+        description = "The path of an existing key file of the root certificate authority.";
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    systemd.services.step-ca = {
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+      path = [ pkgs.step-ca pkgs.step-cli ];
+      environment = { STEPPATH = "/var/lib/step-ca"; HOME = "/var/lib/step-ca"; };
+      script = ''
+        if [ ! -f "$STATE_DIRECTORY/ca-is-init" ]; then
+          step ca init \
+            --ssh \
+            --name ${cfg.name} \
+            --dns ${cfg.hostname} \
+            --address ${cfg.address}:${toString cfg.port} \
+            --provisioner ${cfg.provisioner} \
+            --password-file ${cfg.passwordFile} \
+            ${lib.optionalString (cfg.rootFile != null) " --root=${cfg.rootFile} "} \
+            ${lib.optionalString (cfg.keyFile != null) " --key=${cfg.keyFile} "} \
+            ${lib.optionalString (cfg.acmeProvisioner != null) "step ca provisioner add ${cfg.acmeProvisioner} --type ACME"}
```

Comment:
I tested that this \ thing works locally but I would like some confirmation.

Reply:
Ok, something doesn't work there - I need to check (likely tomorrow)

Reply:
Found it, one \ too much.

Reply:
One issue here is that the `--password-file` when passed to `step ca init` will be used both for the first (JWK) provisioner password, and the CA private key password.

When `step-ca --password-file` is called, the password expected for that parameter is the intermediate CA private key password only. (The JWK provisioner password is supplied by the user when they use that provisioner.)

It is recommended that the user do one of the following:
* Supply a different `--provisioner-password-file` to `step ca init` (this is probably the option that makes the most sense for you here)
* Change the CA private key password after initializing, using `step certificate change-pass $(step path)/secrets/intermediate_ca_key`
* Change the JWK provisioner password after initializing the CA, so that it's not the same as the private key pw.
* Or, don't create a JWK provisioner at all and use a different type (eg. ACME) as the first provisioner. The JWK provisioner is easy for admins to use to make one-off certificates as needed, but there are better options for any sort of automation.

[See this discussion for more context](https://github.com/smallstep/certificates/discussions/427)

Reply:
Also, Moritz, I'm happy to jump onto a video chat with you if you want to go through any of this together. You can reach me at carl at smallstep.com.

Reply:
Thanks a lot for the offer @tashian! I may get on Gitter sometime (likely not today as I don't have that much time (I'm studying CS)) but I try to keep most of the relevant and general discussion here and would only discuss the details there so other people could find the information easily later if they want to extend this module or fix a bug etc.